### PR TITLE
*: rewrite ifnull

### LINKED ISSUE #4047: incorrect result: aggregation function `sum`
## 1. What did you do?
```sql
use test;
drop table if exists t1;
drop table if exists t2;
create table t1(a decimal(20,4));
create table t2(a decimal(20,4));

insert into t1 select 1.2345;
insert into t2 select 1.2345;

select sum(ifnull(a, 0)) from (
    select ifnull(a, 0) as a from t1
    union all
    select ifnull(a, 0) as a from t2
) t;
```

## 2. What did you expect to see?

```sql
+-------------------+
| sum(ifnull(a, 0)) |
+-------------------+
|            2.4690 |
+-------------------+
1 row in set (0.00 sec)
```

## 3. What did you see instead?

```sql
+-------------------+
| sum(ifnull(a, 0)) |
+-------------------+
|                 2 |
+-------------------+
```

## 4. What version of TiDB are you using (`tidb-server -V`)?

6e0ff81269d49ffec39066d5110ee28f7fe67f2f

### PR DESCRIPTION
**Pull request by XuHuaiyu**
fix #4047 
PTAL @zz-jason @shenli

### CODE REVIEW

File: util/types/field_type.go [hunk 1]
```diff
@@ -62,6 +62,69 @@ func NewFieldType(tp byte) *FieldType {
 	}
 }
 
+// AggFieldType aggregates field types.
```
Comment (shenli) [inline util/types/field_type.go]: Any test case covers the following changes?

Comment (shenli) [inline util/types/field_type.go]: The comment is not clear.

File: expression/builtin_control_test.go [hunk 2]
```diff
@@ -81,23 +81,39 @@ func (s *testEvaluatorSuite) TestIf(c *C) {
...
+		arg1     interface{}
+		arg2     interface{}
+		expected interface{}
+		isNil    bool
+		getErr   bool
 	}{
-		{1, 2, 1},
-		{nil, 2, 2},
-		{nil, nil, nil},
+		{1, 2, int64(1), false, false},
```
Comment (shenli) [inline expression/builtin_control_test.go]: Add more test cases to cover more types.

File: expression/builtin_control.go [hunk 3]
```diff
@@ -125,32 +130,102 @@ type ifNullFunctionClass struct {
...
-// eval evals a builtinIfNullSig.
-// See https://dev.mysql.com/doc/refman/5.7/en/control-flow-functions.html#function_ifnull
-func (b *builtinIfNullSig) eval(row []types.Datum) (types.Datum, error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+func (b *builtinIfNullIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalInt(row, sc)
+	if !isNull {
```
Comment (zz-jason) [inline expression/builtin_control.go]: check `err` before check `isNull`

File: expression/builtin_control.go [hunk 4]
```diff
@@ -125,32 +130,102 @@ type ifNullFunctionClass struct {
...
+		return arg0, false, errors.Trace(err)
 	}
-	// ifnull(expr1, expr2)
-	// if expr1 is not null, return expr1, otherwise, return expr2
-	v1 := args[0]
-	v2 := args[1]
+	return b.args[1].EvalInt(row, sc)
```
Comment (zz-jason) [inline expression/builtin_control.go]: if `err != nil`, maybe we should return `0, true, errors.Trace(err)`  internally in this function?

File: expression/builtin_control.go (hunk 3 above)
Comment (XuHuaiyu) [inline expression/builtin_control.go]: we do not check err here,
we return errors.Trace(err) in next line,
if err == nil , errors.Trace(err) will be null.

File: expression/builtin_control.go [hunk 5]
```diff
@@ -29,8 +31,11 @@ var (
 var (
 	_ builtinFunc = &builtinCaseWhenSig{}
 	_ builtinFunc = &builtinIfSig{}
-	_ builtinFunc = &builtinIfNullSig{}
 	_ builtinFunc = &builtinNullIfSig{}
+	_ builtinFunc = &builtinIfNullIntSig{}
+	_ builtinFunc = &builtinIfNullRealSig{}
+	_ builtinFunc = &builtinIfNullDecimalSig{}
+	_ builtinFunc = &builtinIfNullStringSig{}
```
Comment (shenli) [inline expression/builtin_control.go]: Add builtinIfNullDurationSig and builtinIfNullTimeSig here.

File: util/types/field_type.go [hunk 6]
```diff
@@ -62,6 +62,69 @@ func NewFieldType(tp byte) *FieldType {
...
+			tpClass = mergeTypeClass(tpClass, argTypeClass, unsigned, mysql.HasUnsignedFlag(argFieldType.Flag))
+			unsigned = unsigned && mysql.HasUnsignedFlag(argFieldType.Flag)
+		}
+	}
+	setTypeFlag(flag, uint(mysql.UnsignedFlag), unsigned)
+	setTypeFlag(flag, uint(mysql.BinaryFlag), tpClass != ClassString || gotBinString)
+	return tpClass
+}
+
+func mergeTypeClass(a, b TypeClass, aUnsigned, bUnsigned bool) TypeClass {
```
Comment (shenli) [inline util/types/field_type.go]: Is this the same work with https://github.com/pingcap/tidb/blob/master/util/types/field_type.go#L306 ？

Comment (XuHuaiyu) [inline util/types/field_type.go]: Noop, [MergeFieldType](https://github.com/pingcap/tidb/blob/master/util/types/field_type.go#L306) does the same work.

Comment (XuHuaiyu) [inline util/types/field_type.go]: mergeTypeClass does the same work with
https://github.com/mysql/mysql-server/blob/5.7/sql/item_cmpfunc.cc#L44

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @shenli

**Comment by shenli** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason

**Pull request merged**